libnetwork/endpoint.go

```diff
@@ -15,6 +15,49 @@ import (
 	"github.com/docker/docker/libnetwork/types"
 )
 
+// ByNetworkType sorts a [Endpoint] slice based on the network-type
+// they're attached to. It implements [sort.Interface] and can be used
+// with [sort.Stable] or [sort.Sort].
+//
+// It is used by [Sandbox.ResolveName] when resolving names in swarm mode.
+// In swarm mode, services with exposed ports are connected to user
+// overlay network, ingress network, and local ("docker_gwbridge")
+// networks.
+//
+// Name resolution should prioritize returning the VIP/IPs on user
+// overlay network over ingress and local networks.
+//
+// ByNetworkType re-order the endpoints, based on the network-type
+// they're attached to:
+//
+// 1. dynamic networks (user overlay networks)
+// 2. ingress network(s)
+// 3. local networks ("docker_gwbridge")
+type ByNetworkType []*Endpoint
+
+func (ep ByNetworkType) Len() int      { return len(ep) }
+func (ep ByNetworkType) Swap(i, j int) { ep[i], ep[j] = ep[j], ep[i] }
+func (ep ByNetworkType) Less(i, j int) bool {
+	return getNetworkType(ep[i].getNetwork()) < getNetworkType(ep[j].getNetwork())
+}
+
+const (
+	typeDynamic = 1
+	typeIngress = 2
+	typeLocal   = 3
+)
```

Comment:
Let me change this to use `iota`, which I think is clearer that the numbers are to important, just their relative value.